Java-Inheritance.md

```diff
@@ -0,0 +1,168 @@
+# Java Inheritance
+
+Java inheritance refers to the ability of a Java Class to _inherit_ the properties from some other Class. Think of it like a child inheriting properties from its parents, the concept is very similar to that. In Java lingo, it is also called _extend_-ing a class. Some simple things to remember :
+
+* The Class that extends or inherits is called **subclass**
+* The Class that is being extended or inherited is called **superclass**
+
+Thus, inheritance gives Java the cool capability of _re-suing_ a code, or share codes between classes! Lets describe it with the classic example of a `Vehicle` class and a `Car` class :
+
+```java
+public class Vehicle {
+	public void start() {
+		// starting the engine
+	}
+
+	public void stop() {
+		// stopping the engine
+	}
+}
+
+public class Car extends Vehicle {
+	int numberOfSeats = 4;
+
+	public int getNumberOfSeats() {
+		return numberOfSeats;
+	}
+}
+```
+
+Thus, here we can see the `Car` class inheriting the properties of the `Vehicle` class. So, we dont have to write the same code of `start()` and `stop()` for Car as well, as those properties comes from its parent. Yes, objects created from the `Car` class _also_ has those properties!
+
+```java
+Car tesla = new Car();
+
+tesla.start();
+
+tesla.stop();
+```
+
+But, does the parent class has the methods of the child? No, it doesn't.
+
+Thus, whenever you need to share some common bit of code to some more classes, it is always good to have a parent Class, and then extend that Class whenever needed! Saves a lot of lines of code, makes code modular and better testable.
+
+### What can be inherited ?
+
+* All `protected` and `public` fields and methods from parent
+
+### What cannot be inherited ?
+
+* `private` fields and methods
+* Constructors. Although, subclass constructor _has_ to call superclass constructor (More on that later!)
+* Multiple classes. Java supports only Single Inheritance, that is you can only inherit one class at a time.
+* Fields. Individual fields of a class cannot be overriden by the subclass.
+
+## Type Casting & Reference
+
+In Java, it is possible to reference a subclass as an _instance_ of its superclass. For example, `Car` class object can be referenced as a `Vehicle` class instance like this : 
+
+```java
+Vehicle car = new Car();
+```
+
+Although, the opposite is not possible :
+
+```java
+Car car = new Vehicle(); // ERROR
+```
+
+So, since you can reference a Java subclass as a superclass instance, you can cast an subclass object easily to a superclass instance. It may be possible to cast a superclass object into a subclass type, but _only if the object is really an instance of subclass_. So keep this in mind :
```

Comment:
> So, since you can reference a Java subclass as a

No need to use _since_ followed by a _So_. Use anyone of the them. But not both in consecution.